File: MEETUPS/Machine_Test_Script.py
#!/usr/bin/env python3
import os, platform, pwd
from io import StringIO
import re

import numpy as np
import pandas as pd
import datetime as dt

import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

from itertools import combinations, chain
import statsmodels.api as sm
import statsmodels.formula.api as smf

import psutil

### SETTINGS
mpl.use('Agg')
plt.ioff()

pd.set_option('display.width', 1000)
pd.set_option('display.max_columns', None) 
pd.set_option('max_colwidth', -1)

#################
### CPU SPECS
#################

sys_data = pd.DataFrame({'user': pwd.getpwuid(os.getuid())[0],
                         'os_type': platform.system(),
                         'os_release': platform.release(),
                         'os_version': platform.platform(),
                         'machine': platform.machine(),
                         'python_version': platform.python_version()
                        }, index=[0])
    
free_data = pd.DataFrame({'mem_total': np.nan, 
                          'mem_used': np.nan,
                          'mem_free': np.nan,
                          'swap_total': np.nan, 
                          'swap_used': np.nan,
                          'swap_free': np.nan
                         }, index=[0])

### LINUX MACHINES
if platform.system() in ["Linux", "Linux2"]:
    try:
        free_raw = tuple(map(int, os.popen('free').readlines()[1].split()[1:])) 
        swap_raw = tuple(map(int, os.popen('free').readlines()[2].split()[1:])) 

        free_data['mem_total'] = free_raw[0] / 1E3 
        free_data['mem_used'] = free_raw[1] / 1E3
        free_data['mem_free'] = free_raw[2] / 1E3
        free_data['swap_total'] = swap_raw[0] / 1E3
        free_data['swap_used'] = swap_raw[1] / 1E3
        free_data['swap_free'] = swap_raw[2] / 1E3

    except Exception as e:
        print(e)

    try:
        free_data['cpu_cores'] = int(os.popen("getconf _NPROCESSORS_ONLN").readlines()[0].split()[0])
    except Exception as e:
        print(e)

    try:
        free_data['cpu_speed'] = float(os.popen("lscpu | grep MHz").readlines()[0].split(':')[1].strip())
    except Exception as e:
        print(e)

    
### MAC OS MACHINES
elif platform.system() == "Darwin":
    try:
        txt = [t for t in os.popen("top -l 1 | head").readlines() if "PhysMem" in t][0].rstrip()
    
        txt = re.sub("PhysMem: | \(|\), ", "", txt)
        txt = re.sub(" used| wired| unused", ",", txt)
        vec = [int(t[:-1])*1000 if t.endswith('G') else int(t[:-1]) for t in txt.split(",")[:-1]]

        free_data['mem_total'] = sum(vec)
        free_data['mem_used'] = vec[0]
        free_data['mem_free'] = vec[2]

    except Exception as e:
        print(e)

    try:
        free_data['cpu_cores'] = os.popen("getconf _NPROCESSORS_ONLN").readlines()[0].rstrip()
    except Exception as e:
        print(e)

    try:
        free_data['cpu_speed'] = float(os.popen("sysctl hw.cpufrequency").readlines()[0].split(':')[1].strip()) / 1E6    
    except Exception as e:
        print(e)

### WINDOWS MACHINES
elif platform.system() in ["win32", "win64"]:
    try:
        free_data = (pd.concat([pd.read_table(StringIO(os.popen("wmic ComputerSystem get TotalPhysicalMemory").readlines()), sep="\s+"),
                                pd.read_table(StringIO(os.popen("wmic OS get FreePhysicalMemory,TotalVirtualMemorySize,FreeVirtualMemory").readlines()), sep="\s+")],
                               axis=1)
                       .rename(columns=["mem_total", "mem_free", "swap_free", "swap_total"])
                    )
    except Exception as e:
        print(e)
  
    try: 
        free_data['cpu_cores'] = os.popen("wmic cpu get NumberOfCores").readlines()[1].strip()
    except Exception as e:
        print(e)

    try:
        free_data['cpu_speed'] = os.popen("wmic cpu get CurrentClockSpeed").readlines()[1].strip()
    except Exception as e:
        print(e)

sys_data = sys_data.join(free_data)


def machine_run():
        
    #################
    ### DATA BUILD
    #################

    alpha = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
    data_tools = ['sas', 'stata', 'spss', 'python', 'r', 'julia']

    np.random.seed(2019)
    random_df = pd.DataFrame({'group': np.random.choice(data_tools, size=int(1E6)),
                              'float': np.random.randint(1, 15, int(1E6)),
                              'y_num': np.random.uniform(0, 100, int(1E6)),
                              'x_num1': np.random.uniform(0, 100, int(1E6)),
                              'x_num2': np.random.uniform(size=int(1E6)) * 10,
                              'x_num3': np.random.normal(10, 1, int(1E6)),
                              'x_num4': np.random.normal(50, 1, int(1E6)),
                              'x_num5': np.random.normal(100, 1, int(1E6)),
                              'char': [''.join(np.random.choice(list(alpha), 3)) for _ in range(int(1E6))],
                              'bool': np.random.choice([True, False], int(1E6)),
                              'date': np.random.choice(pd.date_range('2000-01-01', 
                                                                     dt.datetime.today().strftime('%Y-%m-%d')), int(1E6)),
                               })
    
    #################
    ### AGGREGATION
    #################
    
    agg_df = (random_df.groupby('group')['y_num', 'x_num1', 'x_num2', 'x_num3', 'x_num4', 'x_num5']
                       .agg(['sum', 'mean', 'median', 'min', 'max'])
             )
    
    
    #################
    ### PLOT
    #################
    
    sns.set()
    
    # HISTOGRAM BY GROUP
    plt.figure(figsize=(15,5))
    
    plt.hist(random_df.query("group == 'julia'")['y_num'], bins=50, label='julia')
    plt.hist(random_df.query("group == 'r'")['y_num'], bins=50, label='r')
    plt.hist(random_df.query("group == 'python'")['y_num'], bins=50, label='python')
    plt.hist(random_df.query("group == 'sas'")['y_num'], bins=50, label='sas')
    plt.hist(random_df.query("group == 'stata'")['y_num'], bins=50, label='stata')
    plt.hist(random_df.query("group == 'spss'")['y_num'], bins=50, label='spss')
    
    plt.title("Overlapping Histograms", weight='bold', size=24)
    plt.legend(loc='upper right')
    plt.clf()
    plt.close('all')
    
    # BAR PLOT BY GROUP AND YEAR
    fig, ax = plt.subplots(figsize=(15,6))
    
    grp_df = (random_df.assign(year = random_df['date'].dt.year)
                       .groupby(['group','year'], as_index=False)['y_num']
                       .mean()
              )
    
    sns.barplot(x='year', y='y_num', hue='group', data=grp_df)
    
    plt.title('Barplot by Group and Year', weight='bold', size=20)
    plt.xlabel('Groups', weight='bold', size=18)
    plt.ylabel('Mean of Y', weight='bold', size=16)
    plt.xticks(ha='right')
    
    plt.tight_layout()
    plt.clf()
    plt.close()
    

    #################
    ### MODEL RUN
    #################
    
    def powerset(iterable):
        "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
        s = list(iterable)
        return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))
    
    expvar_list = filter(len, powerset(['x_num'+str(i) for i in range(1,6)]))
    formulas_list = ['y_num ~ {} + group'.format(" + ".join(f)) for f in list(expvar_list)]
            
    models = [smf.ols(f, data=sm.add_constant(random_df)).fit() for f in formulas_list]
    
    
    #################
    ### RESULTS
    #################
    
    coeff_dfs = [pd.DataFrame({'coeff': m.params, 
                               't_stats': m.tvalues,
                               'std_err': m.bse,
                               'p_values': m.pvalues,
                               'low_ci': m.conf_int()[0],
                               'high_ci': m.conf_int()[1]}) for m in models]
    
    anova_dfs = [pd.DataFrame({'obs': m.nobs,
                               'df_resid': m.df_resid,
                               'r_sq': m.rsquared,
                               'r_sq_adj': m.rsquared_adj,
                               'f_value': m.fvalue,
                               'f_pvalue': m.f_pvalue,
                               'ssr': m.ssr,
                               'ess': m.ess,
                               'mse_model': m.mse_model,
                               'mse_resid': m.mse_resid,
                               'mse_total': m.mse_total}, index=[0]) for m in models]
    
    plot_results = []
    
    for m in models:
        fig, ax = plt.subplots(figsize=(15,5))
            
        sns.barplot(x=m.params[1:].index, y=m.params[1:])
        
        plt.title('Barplot of Coefficients', weight='bold', size=18)
        plt.xlabel('Explanatory Variables', weight='bold', size=16)
        plt.ylabel('Estimate', weight='bold', size=16)
        plt.xticks(ha='center')
        
        plt.tight_layout()
        plt.clf()
        plt.close()
        
        plot_results.append(fig)
    

#####################
### BENCHMARK
#####################

if __name__=='__main__':
    from timeit import Timer

    ### RUNTIME MEMORY    
    machine_run()

    # NUMBERS IN BYTES
    process = psutil.Process(os.getpid()).memory_info()      
    sys_data = sys_data.join(pd.DataFrame({'physical_memory_used': process.rss / 1E6, 
                                           'virtual_memory_used': process.vms / 1E6
                                          }, index=[0])
                            )
 
    # TIMINGS
    f = Timer("machine_run()", "from __main__ import machine_run")
    res = f.repeat(repeat=5, number=1)
    
    sys_data = sys_data.join(pd.DataFrame({'min': np.min(res),
                                           'mean': np.mean(res),
                                           'median': np.median(res),
                                           'max': np.max(res),
                                           'neval': 5}, index=[0])
                            )

# OUTPUT TO SCREEN
print(sys_data)

# OUTPUT TO FILE
sys_data.to_csv("Machine_Test_Results_py.csv", index=False)

File: MEETUPS/CRUG/Machine_Test_Script.R

library(microbenchmark)

#################
### CPU SPECS
#################

sys_data <- data.frame(user = Sys.info()['user'],
                       os_type = Sys.info()['sysname'],
                       os_release = Sys.info()['release'],
                       os_version = Sys.info()['version'],
                       machine = Sys.info()['machine'],
                       r_version = paste(R.Version()$major, R.Version()$minor, sep="."),
                       r_platform =  R.Version()$platform,
                       r_build_date = paste(R.Version()$year, R.Version()$month, R.Version()$day, sep="-"))


free_data <- data.frame(mem_total = NA, mem_used = NA, mem_free = NA,
                        swap_total = NA, swap_used = NA, swap_free = NA,
                        cpu_cores = NA, cpu_speed = NA)

### LINUX MACHINES
if(Sys.info()[['sysname']] == "Linux") {
  tryCatch({
      free <- system("free", intern=TRUE)
      free[3] <- paste(free[3], " NA NA NA")
  
      free_data <- read.table(text=paste(free, collapse="\n"), header=TRUE)[1:3]
      free_data <- setNames(cbind(free_data["Mem",], free_data["Swap",],  row.names=NULL) / 1E3, 
                        c("mem_total", "mem_used", "mem_free", "swap_total", "swap_used", "swap_free"))
  }, error = function(e) print(e) )

  tryCatch({ free_data$cpu_cores <- system("getconf _NPROCESSORS_ONLN", intern=TRUE) }, 
           error = function(e) print(e) )

  tryCatch({ free_data$cpu_speed <- trimws(strsplit(system("lscpu | grep MHz", intern=TRUE)[1], ":")[[1]][2]) }, 
           error = function(e) print(e) )
}


### MAC OS X MACHINES
if(Sys.info()[['sysname']] == "Darwin") {
  tryCatch({
      txt <- system("top -l 1 | head -n 10", intern=TRUE)
      txt <- txt[grepl("PhysMem", txt)]

      ignore_txt <- strsplit(txt, "([0-9])+G")[[1]]
      for (w in ignore_txt) {txt <- gsub(w, "", txt, fixed = TRUE)}
      vec <- as.numeric(strsplit(txt, "G")[[1]]) * 1000
      free_data <- data.frame(mem_total = sum(vec), mem_used = vec[1], mem_free = vec[2])
  }, error = function(e) print(e) )

  tryCatch({ free_data$cpu_cores <- system("getconf _NPROCESSORS_ONLN", intern=TRUE) }, 
           error = function(e) print(e) )

  tryCatch({ free_data$cpu_speed <- as.numeric(strsplit(system("sysctl hw.cpufrequency", intern=TRUE), ":")[[1]][2]) / 1E6 }, 
           error = function(e) print(e) )
}


### WINDOWS MACHINES
if(Sys.info()[['sysname']] == "Windows") {
  tryCatch({
      free_data <- setNames(cbind(read.table(text=system("wmic ComputerSystem get TotalPhysicalMemory", intern = TRUE),
                                             header=TRUE),
                                  read.table(text=system("wmic OS get FreePhysicalMemory,TotalVirtualMemorySize,FreeVirtualMemory", intern = TRUE),
                                             header=TRUE)),
                            c("mem_total", "mem_free", "swap_free", "swap_total"))
    
      free_data <- within(free_data, {
		        mem_total <- mem_total / 1E6
		        mem_free <- mem_free / 1E3
		        mem_used <- mem_total - mem_free
		        swap_total <- swap_total / 1E3
		        swap_free <- swap_free / 1E3		
		        swap_used <- swap_total - swap_free
      })
			
      free_data <- free_data[c("mem_total", "mem_used", "mem_free", "swap_total", "swap_used", "swap_free")] 
  }, error = function(e) print(e) )
  
  tryCatch({ free_data$cpu_cores <- trimws(system("wmic cpu get NumberOfCores", intern=TRUE))[2] },
           error = function(e) print(e) )

  tryCatch({ free_data$cpu_speed <- trimws(system("wmic cpu get CurrentClockSpeed", intern=TRUE))[2] },
           error = function(e) print(e) )
}

sys_data <- cbind(sys_data, free_data)


machine_run <- function() {
  
  #################
  ### DATA BUILD
  #################
  
  set.seed(2019)
  alpha <- c(LETTERS, letters, c(0:9))
  data_tools <- c("sas", "stata", "spss", "python", "r", "julia")
  
  random_df <- data.frame(
    group = sample(data_tools, 500, replace=TRUE),
    int = sample(1:15, 1E4, replace=TRUE),
    y_num = runif(1E4, 0, 100),
    x_num1 = runif(1E4, 0, 100),
    x_num2 = runif(1E4) * 10,
    x_num3 = rnorm(1E4, 10),
    x_num4 = rnorm(1E4, 50),
    x_num5 = rnorm(1E4, 100),
    char = replicate(1E4, paste(sample(alpha, 3, replace=TRUE), collapse="")),
    bool = sample(c(TRUE, FALSE), 1E4, replace=TRUE),
    date = as.Date(sample(as.integer(as.Date("2000-01-01")):as.integer(Sys.Date()), 
                          1E4, replace=TRUE), origin="1970-01-01")
  )

  #################
  ### AGGREGATION
  #################
  
  agg_df <- do.call(data.frame, 
                    aggregate(cbind(y_num, x_num1, x_num2, x_num3, x_num4, x_num5) ~ group, random_df, 
                              FUN=function(x) c(sum=sum(x), 
                                                mean=mean(x),
                                                median=median(x),
                                                min=min(x),
                                                max=max(x)))
  )
  
  agg_df <- cbind(agg_df[c(1, grep("sum", names(agg_df)))],
                  agg_df[grep("mean", names(agg_df))],
                  agg_df[grep("median", names(agg_df))],
                  agg_df[grep("min", names(agg_df))],
                  agg_df[grep("max", names(agg_df))])
  
  agg_df 
  
  
  #################
  ### PLOT
  #################
  
  seaborn_palette <- c('#4878d0', '#ee854a', '#6acc64', '#d65f5f', '#956cb4', '#8c613c', 
                       '#dc7ec0', '#797979', '#d5bb67', '#82c6e2', '#4878d0', '#ee854a', 
                       '#6acc64', '#d65f5f', '#956cb4', '#8c613c', '#dc7ec0', '#797979', 
                       '#d5bb67', '#82c6e2', '#4878d0', '#ee854a', '#6acc64', '#d65f5f')
  
  # HISTOGRAM BY GROUP
  pdf(NULL)
  hist(subset(random_df, random_df$group == "julia")$y_num, 
       col=rgb(0,0,1,0.25), breaks=50, ylim=c(0, 5000),
       main="Overlapping Histogram", xlab="y_num")
  hist(subset(random_df, random_df$group == "r")$y_num, col=rgb(0,0,1,0.5), breaks=50, add=TRUE)
  hist(subset(random_df, random_df$group == "python")$y_num, col=rgb(0,0,1,0.75), breaks=50, add=TRUE)
  hist(subset(random_df, random_df$group == "sas")$y_num, col=rgb(1,0,0,0.5), breaks=50, add=TRUE)
  hist(subset(random_df, random_df$group == "stata")$y_num, col=rgb(1,0,0,0.75), breaks=50, add=TRUE)
  hist(subset(random_df, random_df$group == "spss")$y_num, col=rgb(1,0,0,0.75), breaks=50, add=TRUE)
  
  legend("top", c("julia", "r", "python", "sas", "stata", "spss"), ncol=6,
         fill=c(rgb(0,0,1,0.25), rgb(0,0,1,0.5), rgb(0,0,1,0.75),
                rgb(1,0,0,0.25), rgb(1,0,0,0.5), rgb(1,0,0,0.75)))
  box()
  dev.off()

  # BAR PLOT BY GROUP AND YEAR
  random_df$year <- format(random_df$date, "%Y")
  graph_data <- tapply(random_df$y_num, list(random_df$group, random_df$year), mean)
  pdf(NULL)
  barplot(graph_data, beside=TRUE, col=seaborn_palette[1:6], legend=NULL, 
          ylim=range(pretty(c(0, graph_data))),
          main="Barplot by Group and Year", ylab="y_num")
  
  legend("top", row.names(graph_data), ncol=6, fill=seaborn_palette[1:6])
  box()
  dev.off()
  
  #################
  ### MODEL RUN
  #################
  
  expvar_list <- lapply(1:5, function(x) combn(paste0("x_num", 1:5), x, simplify=FALSE))
  
  formulas_list <- rapply(expvar_list, function(x) paste(paste("y_num ~", paste(x, collapse=" + ")), " + group"))
  formulas_list
  
  models <- lapply(formulas_list, function(m) lm(m, data=random_df))
  
  #################
  ### RESULTS
  #################
  
  coeff_matrices <- vector(mode="list", length = length(models))
  rf_matrices <- vector(mode="list", length = length(models))
  anova_dfs <- vector(mode="list", length = length(models))
  plot_results  <- vector(mode="list", length = length(models))
  
  build_plot <-  function(i){
    pdf(NULL)
    dev.control(displaylist="enable")
    barplot(coeff_matrices[[i]][,1], col=seaborn_palette[1:nrow(coeff_matrices[[i]])],
            legend = row.names(coeff_matrices[[i]]), ylim=c(-10, 60),
            main=paste0("Model - ", i))
    box()
    
    p <- recordPlot()
    invisible(dev.off())
    return(p)
  }
  
  for(i in seq_along(models)){
    res <- summary(models[[i]])
    
    coeff_matrices[[i]] <- res$coefficients
    
    p <- res$fstatistic
    rf_matrices[[i]] <- c(r_sq = res$r.squared,
                          adj_r_sq = res$adj.r.squared,
                          f_stat = p[['value']],
                          p_value = unname(pf(p[1], p[2], p[3], lower.tail=FALSE))
    )
    
    anova_dfs[[i]] <- anova(models[[i]])
    
    plot_results[[i]] <- build_plot(i)
  }
}

#####################
### RUNTIME MEMORY
#####################

machine_run()
mem_stats <- gc()
colnames(mem_stats) <- c("used", "used_mb", "gc_trigger", "gc_trigger_mb", "max_used", "max_used_mb")
mem_stats <- colSums(mem_stats)

sys_data <- within(sys_data, { 
                use_mb <- mem_stats['used_mb']
                gc_trigger_mb <- mem_stats['gc_trigger_mb']
                max_used_mb <- mem_stats['max_used_mb']
            })


#####################
### BENCHMARK
#####################

mb_stats <- do.call(data.frame, aggregate(time~., microbenchmark(machine_run(), times=5L, unit="s"), function(x)
  c(min = min(x) / 1E9,
    lq = unname(quantile(x)[2]) / 1E9,
    mean = mean(x) / 1E9,
    median = median(x) / 1E9,
    uq = unname(quantile(x)[4]) / 1E9,
    max = max(x) / 1E9,
    neval = 5L)
))


sys_data <- cbind(sys_data, mb_stats, row.names=NULL)

# OUTPUT TO SCREEN
print(sys_data)

# OUTPUT TO FILE
write.csv(sys_data, "Machine_Test_Results_r.csv", row.names=FALSE)
